### PR TITLE
wdl-analysis: prefer doc comments over meta for LSP docs

### LINKED ISSUE #569: refactor: move display methods out of `wdl-ast`
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is.

This TODO has been around for a long time - https://github.com/stjude-rust-labs/sprocket/blob/8b1b848f6eb345c1d83dbae5d207272a5512f9cd/crates/wdl-ast/src/v1/display.rs#L1-L5

`wdl-ast` shouldn't be concerned with rendering AST elements as markdown.

**Describe the solution you'd like**

I'm not even sure this makes that much sense to belong in `wdl-analysis` (as the TODO suggests), but that would be the easiest way for `wdl-doc` and `wdl-lsp` to share their functionality.

**Describe alternatives you've considered**

I haven't reviewed these parts of the codebase closely in recent memory, so I have not considered alternatives.

**Additional context**

Logging this as a tracking issue that should be investigated closer.

### LINKED ISSUE #646: Support doc comments in `wdl-lsp`/`wdl-analysis`
**Is your feature request related to a problem? Please describe.**
Right now, only the traditional `meta`/`parameter_meta` sections are used for hover/completion docs.

**Describe the solution you'd like**
When generating the markdown descriptions, the items should first be checked for doc comments, with `meta`/`parameter_meta` only being used as a fallback.

Additionally, we should take this opportunity to remove the `<Definition>::markdown_description()` methods from `wdl-ast` and move them into `wdl-analysis/src/handlers/common/docs.rs`, where they actually get used.

### PR DESCRIPTION
**Pull request by kris70lesgo**
_Describe the problem or feature in addition to a link to the issues._
This PR updates hover and completion documentation to prefer doc comments over meta / parameter_meta, with existing behaviour preserved as a fallback when no doc comments are present.
Markdown rendering logic has been moved out of wdl-ast into wdl-analysis, where it is consumed by the LSP handlers. Variable level hover documentation now also checks for doc comments on declarations before falling back to parameter_meta.
Tests cover precedence and fallback behaviour for hover documentation.

Fixes #646.
Closes #569.

Before submitting this PR, please make sure:

For external contributors:

- [x] You have read the [contributing guide](https://github.com/stjude-rust-labs/sprocket/blob/main/CONTRIBUTING.md) in its entirety.
- [x] You have not used AI on any parts of this pull request.

For all contributors:

- [x] You have added a few sentences describing the PR here.
- [x] You have added yourself or the appropriate individual as the assignee.
- [x] You have added at least one relevant code reviewer to the PR.
- [x] Your code builds clean without any errors or warnings.
- [x] You have added tests (when appropriate).
- [ ] You have added an entry in the CHANGELOG (when appropriate).
- [ ] You have updated the README or other documentation to account for these changes (when appropriate).
- [ ] You have made a PR to the `next` branch in the [sprocket.bio repository](https://github.com/stjude-rust-labs/sprocket.bio) (when appropriate).

For PRs containing lint rule changes:

- [ ] You have updated any and all effected entries within `RULES.md`.
- [ ] You have added a test case in `crates/wdl-lint/tests/lints` that covers every
      possible diagnostic emitted for the rule within the file where the rule
      is implemented.

### CODE REVIEW

File: crates/wdl-ast/src/v1.rs [hunk 1]
```diff
@@ -6,7 +6,7 @@ use crate::SyntaxNode;
 use crate::TreeNode;
 
 mod decls;
-mod display;
+pub mod display;
```
Comment (claymcleod) [inline crates/wdl-ast/src/v1.rs]: Would you mind narrowing the visibility here? Making the entire `display` module `pub` exposes internal helpers (e.g., `write_meta_object`, `write_meta_array`) as public API, which creates a semver hazard. Consider re-exporting only the four functions `wdl-analysis` actually needs:

```suggestion
pub(crate) mod display;
pub use display::write_input_section;
pub use display::write_output_section;
pub use display::get_param_meta;
pub use display::format_meta_value;
```

File: crates/wdl-analysis/src/handlers/common/docs.rs [hunk 2]
```diff
@@ -24,6 +36,152 @@ pub fn make_md_docs(definition: String) -> Option<Documentation> {
...
+        return None;
+    }
+
+    let mut lines = Vec::new();
+    let mut current = node.prev_sibling_or_token();
+
+    while let Some(sibling) = current {
+        match sibling {
+            rowan::NodeOrToken::Token(ref token) => {
+                if token.kind() == SyntaxKind::Whitespace {
```
Comment (claymcleod) [inline crates/wdl-analysis/src/handlers/common/docs.rs]: Is it intentional that `extract_doc_comment` skips all whitespace tokens, including blank lines? Two `##` blocks separated by a blank line would be merged into a single doc comment. If this is the intended behavior, consider adding a test case that exercises this scenario (I do see the mention of ignoring whitespace splitting, but the effect I describe above might still be surprising). If not intended, consider breaking the collection when the whitespace contains `\n\n`.

File: crates/wdl-analysis/src/handlers/common/docs.rs [hunk 3]
```diff
@@ -24,6 +36,152 @@ pub fn make_md_docs(definition: String) -> Option<Documentation> {
...
+    } else {
+        None
+    }
+}
+
+/// Renders markdown documentation for a task definition.
+///
+/// Doc comments are preferred over `meta.description` for the description
+/// paragraph. Input and output sections always use `parameter_meta`.
+fn render_task_doc(n: &TaskDefinition, syntax: &SyntaxNode) -> String {
```
Comment (claymcleod) [inline crates/wdl-analysis/src/handlers/common/docs.rs]: `render_task_doc` and `render_workflow_doc` share the same structure (header → doc comment or meta fallback → input/output sections). Consider extracting a shared helper to avoid the duplication — something like `render_runnable_doc(keyword, name, syntax, metadata, input, output, param_meta)`.

File: crates/wdl-analysis/src/handlers/common/docs.rs [hunk 4]
```diff
@@ -24,6 +36,152 @@ pub fn make_md_docs(definition: String) -> Option<Documentation> {
...
+            rowan::NodeOrToken::Token(ref token) => {
+                if token.kind() == SyntaxKind::Whitespace {
+                    current = sibling.prev_sibling_or_token();
+                    continue;
+                }
+                if let Some(comment) = Comment::cast(token.clone())
+                    && comment.is_doc_comment()
+                {
+                    let text = comment.inner().text();
+                    let stripped = text
```
Comment (claymcleod) [inline crates/wdl-analysis/src/handlers/common/docs.rs]: The nested `unwrap_or` reads inside-out and is a bit hard to follow. Consider:

```suggestion
                    let stripped = text
                        .strip_prefix("## ")
                        .or_else(|| text.strip_prefix("##"))
                        .unwrap_or(text);
```

File: crates/wdl-lsp/tests/hover.rs [hunk 5]
```diff
@@ -197,3 +197,64 @@ async fn should_hover_enum_variant() {
...
+#[tokio::test]
+async fn should_hover_task_doc_comment_over_meta() {
+    let mut ctx = setup().await;
+    // Position of `doc_and_meta` in `task doc_and_meta`
+    let response = hover_request(&mut ctx, "doc_comments.wdl", Position::new(20, 7)).await;
+    assert_hover_content(&response, "task doc_and_meta");
+    assert_hover_content(&response, "This doc comment should win over meta.");
+}
+
+fn assert_hover_not_content(hover: &Option<Hover>, unexpected: &str) {
```
Comment (claymcleod) [inline crates/wdl-lsp/tests/hover.rs:222]: `assert_hover_not_content` silently returns when `hover` is `None`, which means a test using it could pass vacuously if the hover response is unexpectedly absent. Consider adding `assert!(hover.is_some())` or `panic!` on `None` to avoid false positives.

File: crates/wdl-lsp/tests/hover.rs [hunk 6]
```diff
@@ -197,3 +197,64 @@ async fn should_hover_enum_variant() {
...
+}
+
+#[tokio::test]
+async fn should_hover_input_doc_comment() {
+    let mut ctx = setup().await;
+    // Position of `name` in `String name` inside doc_only task input
+    let response = hover_request(&mut ctx, "doc_comments.wdl", Position::new(7, 15)).await;
+    assert_hover_content(&response, "(variable) name: String");
+    assert_hover_content(&response, "The person's name to greet");
+}
```
Comment (claymcleod) [inline crates/wdl-lsp/tests/hover.rs:260]: The new tests thoroughly cover task doc comments (precedence, fallback, negative assertion, input-level). Would you consider adding at least one test each for workflow, struct, and enum doc comments? Those `render_*_doc` paths have slightly different logic (e.g., `render_struct_doc` uses `n.metadata().next()`, `render_enum_doc` has no meta fallback) and would benefit from test coverage.

File: crates/wdl-analysis/src/handlers/common/docs.rs (hunk 2 above)
Comment (Serial-ATA) [reply crates/wdl-analysis/src/handlers/common/docs.rs]: ~~@kris70lesgo for this you should wait till #659 is in and switch over to that instead~~

It's merged now. You can get doc comments from `wdl-ast` node types directly

File: crates/wdl-analysis/src/handlers/common/docs.rs [hunk 7]
```diff
@@ -24,6 +36,152 @@ pub fn make_md_docs(definition: String) -> Option<Documentation> {
...
+    }
+
+    let members: Vec<_> = n.members().collect();
+    if !members.is_empty() {
+        let _ = writeln!(s, "\n**Members**");
+        for member in members {
+            let name = member.name();
+            let _ = write!(s, "- **{}**: `{}`", name.text(), member.ty().inner().text());
+            if let Some(meta_val) =
+                get_param_meta(name.text(), n.parameter_metadata().next().as_ref())
```
Comment (Serial-ATA) [inline crates/wdl-analysis/src/handlers/common/docs.rs]: Doc comments can appear on struct members as well. In that case though, you should only render the first paragraph

File: crates/wdl-analysis/src/handlers/common/docs.rs [hunk 8]
```diff
@@ -24,6 +36,152 @@ pub fn make_md_docs(definition: String) -> Option<Documentation> {
...
+    let mut s = String::new();
+    let _ = writeln!(s, "```wdl\nworkflow {}\n```\n---", n.name().text());
+
+    let description = extract_doc_comment(syntax).or_else(|| read_meta_description(n.metadata()));
+    if let Some(desc) = description {
+        let _ = writeln!(s, "{}\n", desc);
+    }
+
+    let _ = write_input_section(&mut s, n.input().as_ref(), n.parameter_metadata().as_ref());
+    let _ = write_output_section(&mut s, n.output().as_ref(), n.parameter_metadata().as_ref());
```
Comment (Serial-ATA) [inline crates/wdl-analysis/src/handlers/common/docs.rs]: Inputs and outputs can also have doc comments

File: crates/wdl-analysis/src/handlers/common/docs.rs [hunk 9]
```diff
@@ -24,6 +36,152 @@ pub fn make_md_docs(definition: String) -> Option<Documentation> {
...
+    let mut s = String::new();
+    let _ = writeln!(s, "```wdl\ntask {}\n```\n---", n.name().text());
+
+    let description = extract_doc_comment(syntax).or_else(|| read_meta_description(n.metadata()));
+    if let Some(desc) = description {
+        let _ = writeln!(s, "{}\n", desc);
+    }
+
+    let _ = write_input_section(&mut s, n.input().as_ref(), n.parameter_metadata().as_ref());
+    let _ = write_output_section(&mut s, n.output().as_ref(), n.parameter_metadata().as_ref());
```
Comment (Serial-ATA) [inline crates/wdl-analysis/src/handlers/common/docs.rs]: Same here

File: crates/wdl-analysis/src/handlers/common/docs.rs [hunk 10]
```diff
@@ -24,6 +37,253 @@ pub fn make_md_docs(definition: String) -> Option<Documentation> {
...
+        return None;
+    }
+    let lines: Vec<String> = comments
+        .iter()
+        .map(|c| {
+            let text = c.inner().text();
+            text.strip_prefix("## ")
+                .or_else(|| text.strip_prefix("##"))
+                .unwrap_or(text)
+                .to_string()
```
Comment (Serial-ATA) [inline crates/wdl-analysis/src/handlers/common/docs.rs]: Rather than assume the comment will be preceded by zero or one spaces, you should just strip the `wdl_ast::DOC_COMMENT_PREFIX` and then determine the minimum common indentation between the lines, like we do in `wdl-doc`:

https://github.com/stjude-rust-labs/sprocket/blob/b32bfbd9c2d6676d2713a5a292528a82ed909c0e/crates/wdl-doc/src/meta.rs#L443-L459

You can just adapt that to iterate over comment lines, rather than paragraphs

File: crates/wdl-analysis/src/handlers/common/docs.rs [hunk 11]
```diff
@@ -24,6 +37,253 @@ pub fn make_md_docs(definition: String) -> Option<Documentation> {
...
+    if comments.is_empty() {
+        return None;
+    }
+    let mut lines: Vec<String> = Vec::new();
+    for c in comments {
+        let text = c.inner().text();
+        let stripped = text
+            .strip_prefix("## ")
+            .or_else(|| text.strip_prefix("##"))
+            .unwrap_or(text);
```
Comment (Serial-ATA) [inline crates/wdl-analysis/src/handlers/common/docs.rs]: Same here

File: crates/wdl-analysis/src/handlers/common/docs.rs [hunk 12]
```diff
@@ -24,6 +37,253 @@ pub fn make_md_docs(definition: String) -> Option<Documentation> {
...
+            Decl::Unbound(u) => first_paragraph_doc(u.doc_comments().unwrap_or_default()),
+            Decl::Bound(b) => first_paragraph_doc(b.doc_comments().unwrap_or_default()),
+        };
+        write!(f, "- **{}**: `{}`", name_text, ty_text)?;
+        if let Some(val) = &default_val {
+            write!(f, " = *`{}`*", val.trim_start_matches(" = "))?;
+        }
+        if let Some(doc_str) = doc {
+            writeln!(f)?;
+            writeln!(f, "  {doc_str}")?;
```
Comment (Serial-ATA) [inline crates/wdl-analysis/src/handlers/common/docs.rs]:
```suggestion
            writeln!(f, "\n")?;
            writeln!(f, "{doc_str}")?;
```

Since this is (potentially) rendering a long paragraph, it might be better to format it so the paragraph is on its own line. What do you think?

File: crates/wdl-analysis/src/handlers/common/docs.rs [hunk 13]
```diff
@@ -24,6 +37,253 @@ pub fn make_md_docs(definition: String) -> Option<Documentation> {
...
+    }
+    writeln!(f, "\n**Outputs**")?;
+    for decl in output.declarations() {
+        let name_text = decl.name().text().to_string();
+        let ty_text = decl.ty().inner().text().to_string();
+        let doc = first_paragraph_doc(decl.doc_comments().unwrap_or_default());
+        write!(f, "- **{}**: `{}`", name_text, ty_text)?;
+        if let Some(doc_str) = doc {
+            writeln!(f)?;
+            writeln!(f, "  {doc_str}")?;
```
Comment (Serial-ATA) [inline crates/wdl-analysis/src/handlers/common/docs.rs]: Same here

File: crates/wdl-analysis/src/handlers/common/docs.rs (hunk 12 above)
Comment (kris70lesgo) [reply crates/wdl-analysis/src/handlers/common/docs.rs]: good point , I will update it so the doc paragraph is rendered on its own line

File: crates/wdl-analysis/src/handlers/common/docs.rs [hunk 14]
```diff
@@ -24,6 +38,265 @@ pub fn make_md_docs(definition: String) -> Option<Documentation> {
...
+/// Converts a list of doc comments to a Markdown string.
+///
+/// Strips [`DOC_COMMENT_PREFIX`] and normalizes indentation. Returns `None` if
+/// the list is empty. All paragraphs are included.
+pub(crate) fn comments_to_string(comments: Vec<Comment>) -> Option<String> {
+    if comments.is_empty() {
+        return None;
+    }
+    let lines = normalize_doc_comments(&comments);
+    Some(lines.join("\n"))
```
Comment (claymcleod) [inline crates/wdl-analysis/src/handlers/common/docs.rs]: `comments_to_string` can return `Some("")` when all doc comments are blank `##` lines—this causes the `or_else` fallback to meta description to be skipped, producing empty hover content instead of the meta description.

```suggestion
    Some(lines.join("\n")).filter(|s| !s.trim().is_empty())
```

File: crates/wdl-analysis/src/handlers/hover.rs [hunk 15]
```diff
@@ -520,9 +523,37 @@ fn get_function_hover_content(name: &str, func: &Function) -> String {
...
+                .and_then(comments_to_string),
+            SyntaxKind::UnboundDeclNode => UnboundDecl::cast(decl_node)
+                .and_then(|d| d.doc_comments())
+                .and_then(comments_to_string),
+            _ => None,
+        };
+        if doc.is_some() {
+            return doc;
+        }
+    }
```
Comment (claymcleod) [inline crates/wdl-analysis/src/handlers/hover.rs]: Consider using `Decl::cast` here—it already handles the `BoundDecl`/`UnboundDecl` branching:

```rust
if let Some(doc) = token.parent_ancestors()
    .find_map(Decl::cast)
    .and_then(|d| match &d {
        Decl::Bound(b) => b.doc_comments(),
        Decl::Unbound(u) => u.doc_comments(),
    })
    .and_then(comments_to_string)
{
    return Some(doc);
}
```

File: crates/wdl-analysis/src/handlers/common/docs.rs [hunk 16]
```diff
@@ -24,6 +38,265 @@ pub fn make_md_docs(definition: String) -> Option<Documentation> {
...
+    if para.is_empty() {
+        None
+    } else {
+        Some(para.join("\n"))
+    }
+}
+
+/// Formats the input section with doc comments preferred over parameter
+/// metadata for each declaration's description.
+fn write_documented_inputs(
```
Comment (claymcleod) [inline crates/wdl-analysis/src/handlers/common/docs.rs]: `write_documented_inputs` and `write_documented_outputs` are nearly identical (same doc-or-meta-or-nothing logic per declaration). Consider unifying into a single generic helper.

**Review by claymcleod — COMMENTED** [review]
A few more things that came up during review.

File: crates/wdl-ast/src/v1.rs [hunk 17]
```diff
@@ -16,6 +16,10 @@ mod tokens;
 mod workflow;
 
 pub use decls::*;
+pub use display::format_meta_value;
+pub use display::get_param_meta;
+pub use display::write_input_section;
+pub use display::write_output_section;
```
Comment (claymcleod) [inline crates/wdl-ast/src/v1.rs]: `write_input_section` and `write_output_section` are re-exported here but no longer consumed outside `wdl-ast` — the callers were removed in this PR. Consider dropping these re-exports to keep the public API surface clean.

```suggestion
pub use display::format_meta_value;
pub use display::get_param_meta;
```
